src/server/saved_objects/migrations/core/elastic_index.ts

```diff
@@ -0,0 +1,284 @@
+
+  return async function read() {
+    const result = await nextBatch();
+    const docs = result.hits.hits;
+
+    scrollId = result._scroll_id;
+
+    if (!docs.length) {
+      await close();
+    }
+
+    return docs;
+  };
+}
+
+/**
+ * Writes the specified documents to the index, throws an exception
+ * if any of the documents fail to save.
+ *
+ * @param {CallCluster} callCluster
+ * @param {string} index
+ * @param {RawDoc[]} docs
+ */
+export async function write(callCluster: CallCluster, index: string, docs: RawDoc[]) {
+  const result = await callCluster('bulk', {
+    body: docs.reduce((acc: object[], doc: RawDoc) => {
+      acc.push({
+        index: {
+          _id: doc._id,
+          _index: index,
+          _type: 'doc',
+        },
+      });
+
+      acc.push(doc._source);
+
+      return acc;
+    }, []),
+  });
+
+  const err = _.find(result.items, 'index.error.reason');
+
+  if (!err) {
+    return;
+  }
+
+  const exception: any = new Error(err.index.error!.reason);
+  exception.detail = err;
+  throw exception;
+}
+
+/**
+ * Checks to see if the specified index is up to date. It does this by checking
+ * that the index has the expected mappings and by counting
+ * the number of documents that have a property which has migrations defined for it,
+ * but which has not had those migrations applied. We don't want to cache the
+ * results of this function (e.g. in context or somewhere), as it is important that
+ * it performs the check *each* time it is called, rather than memoizing itself,
+ * as this is used to determine if migrations are complete.
+ *
+ * @param {CallCluster} callCluster
+ * @param {string} index
+ * @param {MigrationVersion} migrationVersion - The latest versions of the migrations
+ */
+export async function hasMigrations(
+  callCluster: CallCluster,
+  index: string,
+  migrationVersion: MigrationVersion
+): Promise<boolean> {
+  const indexInfo = await fetchInfo(callCluster, index);
+
+  if (!_.get(indexInfo, 'mappings.doc.properties.migrationVersion')) {
+    return false;
+  }
+
+  // If no migrations are actually defined, we're up to date!
+  if (Object.keys(migrationVersion).length <= 0) {
+    return true;
+  }
+
+  const { count } = await callCluster('count', {
+    body: {
+      query: {
+        bool: {
+          should: Object.entries(migrationVersion).map(([type, latestVersion]) => ({
+            bool: {
+              must: [
+                { exists: { field: type } },
+                { bool: { must_not: { term: { [`migrationVersion.${type}`]: latestVersion } } } },
+              ],
+            },
+          })),
+        },
+      },
+    },
+    index,
+    type: 'doc',
+  });
+
+  return count === 0;
+}
+
+/**
+ * Applies the specified mappings to the index.
+ *
+ * @param {CallCluster} callCluster
+ * @param {string} index
+ * @param {IndexMapping} mappings
+ */
+export function putMappings(callCluster: CallCluster, index: string, mappings: IndexMapping) {
```

Comment:
This is true. Most of these functions are tested by the integration test, since they're light on logic, and are basically light-weight wrappers around calls to ElasticSearch, but some of them involve trickier logic, so I unit tested those. Happy to add unit tests for the entire thing, though. I'm just not sure how big a value there is in it.

Reply:
I think we should unit test them. `putMappings` is probably not the best example here, but if you look further down you get things like `convertToAlias` which ultimately results in 5 different request to ES with each one building the last. I can't even tell which integration tests would cover those details necessarily because the integration test is so far removed from this logic.

Reply:
I'm happy to add unit tests. The convertToAlias runs on each migration, so every integration test touches it, and each path should be covered, but I agree that's a somewhat complex function.